### PR TITLE
make MatrixTable.group_by.aggregate_rows/cols work with implicit joins

### PR DESCRIPTION
**Pull request by jbloom22**
Joins were not being tested and fail with source mismatch if joins are present in both key and agg expressions. This fix is analogous to that on Table.group_by.aggregate in #3730, processing all joins at once, and not reprocessing later. I don't address here a deeper bug that throws a source error when processing more than one entry-indexed. I've made an issue #3763

### CODE REVIEW

**Comment by jbloom22** [comment]
@tpoterba here's another wedding gift for when you're back from that trip :)

**Comment by tpoterba** [comment]
punting until tomorrow to get context in person.

**Comment by tpoterba** [comment]
let's make time to talk about this today

File: python/hail/table.py [hunk 1]
```diff
@@ -399,9 +398,32 @@ def _force_count(self):
...
+    def _select_scala(self, row, preserved_key, preserved_key_new, new_key):
+        jt = self._jt
+        if self.key is not None and preserved_key != list(self.key):
+            jt = jt.keyBy(preserved_key)
+        jt = jt.select(row._ast.to_hql(), preserved_key_new, len(preserved_key) if preserved_key else None)
+        if new_key != preserved_key_new:
+            jt = jt.keyBy(new_key)
+        return Table(jt)
+
+    def _make_row(self,
```
Comment (tpoterba) [inline python/hail/table.py]: fix this -- shouldn't be typecheckers.

Also add return type annotation.

```
(might need to from typing import *)
-> Tuple[StructExpression, Tuple[str], ...]
```
```

File: python/hail/matrixtable.py [hunk 2]
```diff
@@ -2885,17 +2874,26 @@ def _select_rows(self, caller, key_struct=None, value_struct=None, pk_size=None)
...
+
+        return cleanup(MatrixTable(base._jvds.selectCols(col._ast.to_hql(), new_key)))
+
+    @typecheck_method(key_struct=nullable(expr_struct()),
+                      value_struct=nullable(expr_struct()))
+    def _select_cols_processed(self, key_struct=None, value_struct=None):
+        col, new_key = self._make_col(key_struct, value_struct)
+        return MatrixTable(self._jvds.selectCols(col._ast.to_hql(), new_key))
+    
+    def _make_col(self, key_struct, value_struct):
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: type annotation

File: python/hail/matrixtable.py [hunk 3]
```diff
@@ -2872,6 +2846,21 @@ def _select_entries(self, caller, s) -> 'MatrixTable':
...
+        return cleanup(MatrixTable(base._jvds.selectRows(row._ast.to_hql(), new_key)))
+
+    @typecheck_method(key_struct=nullable(expr_struct()),
+                      value_struct=nullable(expr_struct()),
+                      pk_size=nullable(int))
+    def _select_rows_processed(self, key_struct=None, value_struct=None, pk_size=None):
+        row, new_key = self._make_row(key_struct, value_struct, pk_size)
+        return MatrixTable(self._jvds.selectRows(row._ast.to_hql(), new_key))
+
+    def _make_row(self, key_struct, value_struct, pk_size):
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: type annotation

File: python/hail/matrixtable.py [hunk 4]
```diff
@@ -314,32 +287,33 @@ def aggregate(self, **named_exprs) -> 'MatrixTable':
...
             if k in fixed_fields:
                 raise ExpressionException("GroupedMatrixTable.aggregate cannot assign duplicate field '{}'".format(k))
             analyze('GroupedMatrixTable.aggregate', v, self._fixed_indices(),
                     {self._parent._row_axis, self._parent._col_axis})
             strs.append('{} = {}'.format(escape_id(k), v._ast.to_hql()))
 
-        base, rename = self._process_keys(base)
+        group_exprs = dict(self._col_keys) if self._col_keys is not None else dict(self._row_keys)
+
+        base, cleanup = self._parent._process_joins(
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: `process_joins(*named_exprs.values(), *group_exprs.values())`

File: python/hail/matrixtable.py [hunk 5]
```diff
@@ -314,32 +287,33 @@ def aggregate(self, **named_exprs) -> 'MatrixTable':
...
+            *itertools.chain(named_exprs.values(), group_exprs.values()))
 
         if self._col_keys is not None:
-            assert self._new_col_keys is not None
-            base = MatrixTable(base.key_cols_by(*self._new_col_keys)._jvds
-                               .aggregateColsByKey(hl.struct(**named_exprs)._ast.to_hql()))
+            keyed_mt = base._select_cols_processed(hl.struct(**dict(self._col_keys)))
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: use group_exprs

File: python/hail/matrixtable.py [hunk 6]
```diff
@@ -314,32 +287,33 @@ def aggregate(self, **named_exprs) -> 'MatrixTable':
...
+            mt = MatrixTable(keyed_mt._jvds.aggregateColsByKey(hl.struct(**named_exprs)._ast.to_hql()))
         elif self._row_keys is not None:
             if self._partition_key is None:
                 self._partition_key = self._row_keys.keys()
-            pk = {k: self._row_keys[k] for k in self._partition_key}
-            rest_of_key = {k: self._row_keys[k] for k in self._row_keys.keys() if k not in self._partition_key}
-            base = MatrixTable(
-                base._key_rows_by("GroupedMatrixTable.aggregate", pk, rest_of_key)
-                ._jvds.aggregateRowsByKey(hl.struct(**named_exprs)._ast.to_hql(), ',\n'.join(strs)))
+            keyed_mt = base._select_rows_processed(hl.struct(**dict(self._row_keys)),
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: use group_exprs

**Pull request merged**